ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/ScalablePushRegistry.java

```diff
@@ -364,4 +457,129 @@ private void runLatest(final LatestConsumer latestConsumerToRun) {
+      catchupConsumer = catchupConsumerFactory.create(ksqlTopic.getKafkaTopicName(), isWindowed(),
+          logicalSchema, consumer, latestConsumer::get, catchupCoordinator, offsetRange,
+          Clock.systemUTC());
+      return catchupConsumer;
+    } catch (Exception e) {
+      LOG.error("Couldn't create catchup consumer", e);
+      processingQueue.onError();
+      // We're not supposed to block here, but if it fails here, hopefully it can immediately close.
+      if (consumer != null) {
+        consumer.close();
+      }
+      throw e;
+    }
+  }
+
+  /**
+   * Starts the catchup consumer
+   * @param processingQueue The queue to register with the catchup consumer.
+   * @param catchupMetadata The catchup metadata
+   */
+  private synchronized void startCatchup(
+      final ProcessingQueue processingQueue,
+      final CatchupMetadata catchupMetadata
+  ) {
+    final CatchupConsumer catchupConsumer = createCatchupConsumer(processingQueue,
+        catchupMetadata.getPushOffsetRange(),
+        catchupMetadata.getCatchupConsumerGroup());
+    catchupConsumer.register(processingQueue);
+    catchupConsumers.put(processingQueue.getQueryId(), catchupConsumer);
+    catchupCounter.incrementAndGet();
+    executorServiceCatchup.submit(() -> runCatchup(catchupConsumer, processingQueue));
+  }
+
+  private void runCatchup(
+      final CatchupConsumer catchupConsumerToRun,
+      final ProcessingQueue processingQueue
+  ) {
+    try (CatchupConsumer catchupConsumer = catchupConsumerToRun) {
+      catchupConsumer.run();
+    } catch (Throwable t) {
+      LOG.error("Got error while running catchup", t);
+      catchupConsumerToRun.onError();
+    } finally {
+      catchupConsumerToRun.unregister(processingQueue);
+      catchupConsumers.remove(processingQueue.getQueryId());
+      catchupCounter.decrementAndGet();
+      // If things ended exceptionally, stop latest
+      stopLatestConsumerOnLastRequest();
+    }
+  }
+
+  public void cleanupCatchupConsumer(final String consumerGroup) {
+    executorServiceCatchup.schedule(() -> {
+      deleteConsumerGroup(consumerGroup);
+    }, 5000L, TimeUnit.MILLISECONDS);
+  }
+
+  public String getCatchupConsumerId(final String suffix) {
+    return sourceApplicationId + CATCHUP_CONSUMER_GROUP_MIDDLE + suffix;
+  }
+
+  private void deleteConsumerGroup(final String consumerGroupId) {
+    LOG.info("Cleaning up consumer group " + consumerGroupId);
+    try {
+      final Map<TopicPartition, OffsetAndMetadata> metadata =
+          serviceContext
+              .getConsumerGroupClient()
+              .listConsumerGroupOffsets(consumerGroupId);
+      if (metadata == null || metadata.values().stream().allMatch(Objects::isNull)) {
+        return;
+      }
+      serviceContext
+          .getConsumerGroupClient()
+          .deleteConsumerGroups(ImmutableSet.of(consumerGroupId));
+    } catch (Throwable t) {
+      LOG.error("Failed to delete consumer group " + consumerGroupId, t);
```

Comment:
Will this lead to resource leakage?

Reply:
bump

Reply:
Well, eventually the consumer groups will be GCed on their own by Kafka.  But yes, they theoretically could.